packages/compartment-mapper/src/types/external.d.ts

```diff
@@ -0,0 +1,352 @@
+
+/**
+ * Options for `importLocation()` necessary (but not sufficient--see
+ * {@link ReadNowPowers}) for dynamic require support
+ */
+export type SyncImportLocationOptions = SyncArchiveOptions & ExecuteOptions;
+
+export type ImportLocationOptions = ArchiveOptions & ExecuteOptions;
+
+// ////////////////////////////////////////////////////////////////////////////////
+// Single Options
+
+type SearchSuffixesOption = {
+  /**
+   * Suffixes to search if the unmodified specifier is not found. Pass `[]` to
+   * emulate Node.js' strict behavior. The default handles Node.js' CommonJS
+   * behavior.
+   * Unlike Node.js, the Compartment Mapper lifts CommonJS up, more like a
+   * bundler, and does not attempt to vary the behavior of resolution depending
+   * on the language of the importing module.
+   */
+  searchSuffixes?: string[];
+};
+
+type SourceMapHookOption = {
+  sourceMapHook?: SourceMapHook;
+};
+
+type ModulesOption = {
+  modules?: Record<string, any>;
+};
+
+type ExitModuleImportHookOption = {
+  importHook?: ExitModuleImportHook;
+};
+
+type ExitModuleImportNowHookOption = {
+  importNowHook?: ExitModuleImportNowHook;
+};
+
+type ParserForLanguageOption = {
+  parserForLanguage?: ParserForLanguage;
+};
+
+type CompartmentOption = {
+  Compartment?: typeof Compartment;
+};
+
+export type ModuleTransformsOption = {
+  moduleTransforms?: ModuleTransforms;
+};
+
+export type SyncModuleTransformsOption = {
+  syncModuleTransforms?: SyncModuleTransforms;
+};
+
+export type PolicyOption = {
+  policy?: any;
+};
+
+type LanguageForExtensionOption = {
+  languageForExtension?: LanguageForExtension;
+};
+
+// ////////////////////////////////////////////////////////////////////////////////
+// Common option groups:
+
+export type SyncOrAsyncArchiveOptions = Partial<{
+  captureSourceLocation: CaptureSourceLocationHook;
+}> &
+  ParserForLanguageOption &
+  CompartmentOption &
+  PolicyOption;
+
+type SyncOrAsyncImportingOptions = SearchSuffixesOption & SourceMapHookOption;
+
+type ImportingOptions = ModulesOption &
+  SyncOrAsyncImportingOptions &
+  ExitModuleImportHookOption;
+
+type SyncImportingOptions = ModulesOption &
+  SyncOrAsyncImportingOptions &
+  ExitModuleImportNowHookOption;
+
+type LinkingOptions = ParserForLanguageOption &
+  CompartmentOption &
+  LanguageForExtensionOption &
+  SyncModuleTransformsOption &
+  ModuleTransformsOption;
+
+// ////////////////////////////////////////////////////////////////////////////////
+
+/**
+ * The result of {@link captureFromMap}.
+ */
+export type CaptureResult = {
+  captureCompartmentMap: CompartmentMapDescriptor;
+  captureSources: Sources;
+  compartmentRenames: Record<string, string>;
+};
+
+/**
+ * The compartment mapper can capture the Sources for all loaded modules
+ * for bundling, archiving, or analysis.
+ */
+export type Sources = Record<string, CompartmentSources>;
+export type CompartmentSources = Record<string, ModuleSource>;
+
+// TODO unionize:
```

Comment:
For a future change, moving from jsdoc substantially improves our chances of turning the `ModuleSource` type from a `Partial` to a union of all valid combinations. That’ll probably complicate the code, but improve our understanding of what properties are expected on module sources of various shapes. They may need a vestigial discriminating property.